### PR TITLE
Test cargo lints

### LINKED ISSUE #5603: Add a way to test cargo lints
There seems to be no way for testing lints in the cargo group currently.

I would be glad to work on this myself!

### PR DESCRIPTION
**Pull request by ebroto**
changelog: Add infrastructure to test cargo lints

Closes #5603

### CODE REVIEW

File: tests/ui/ptr_offset_with_cast.fixed [hunk 1]
```diff
@@ -9,12 +9,12 @@ fn main() {
     let offset_isize = 1_isize;
 
     unsafe {
-        ptr.add(offset_usize);
-        ptr.offset(offset_isize as isize);
-        ptr.offset(offset_u8 as isize);
+        let _ = ptr.add(offset_usize);
```
Comment (flip1995) [inline tests/ui/ptr_offset_with_cast.fixed:12]: This was already added in #5621

**Comment by ebroto** [comment]
@flip1995 I think the test you mention is a good idea! But I've been thinking about a limitation of this implementation, that we could need in some cases to have multiple fail/pass manifests, for example if we want to test multiple values for the same attribute.

I plan to work on #5041 soon, I think it could be a good excuse to try to come up with a solution to the limitation I mentioned. Since this will probably have an impact in the filesystem layout, I think I could implement the test in that PR. 

What do you think?

**Comment by flip1995** [comment]
>  that we could need in some cases to have multiple fail/pass manifests

Instead of checking `[pass, fail]`, you could instead check for `starts_with({pass,fail})`, so directories `fail-this-test` will assume that the cargo lint will fail and `pass-this-test` will assume that the test passes.

On the other hand, why would they even have to be named `pass` or `fail`? Couldn't you just walk every subdir of the cargo tests? A missing `stderr` file would automatically indicate, that a test is pass.

**Comment by ebroto** [comment]
> Couldn't you just walk every subdir of the cargo tests? A missing stderr file would automatically indicate, that a test is pass.

Indeed, that's much simpler, I like it :+1: I will make the changes needed and rework a bit the documentation.

Thanks!

**Comment by ebroto** [comment]
So I've applied the discussed changes. I went for leaving the fail/pass directories by default because I think it's good to enforce having at least one of each cases and IMO it's easier to understand for most of the cases.

I think the test is not viable anymore as the nomenclature is not enforced, let me know if you think otherwise.

**Comment by flip1995** [comment]
Having `pass`/`fail` as default is good.

> I think the test is not viable anymore as the nomenclature is not enforced, let me know if you think otherwise.

Yes, with this, we won't need the test.

**Comment by flip1995** [comment]
Thanks!

@bors r+

**Pull request merged**